[Xamarin.Build.Download] Fix potential race condition while reading/writing assemblies

Redth opened the pull request:
We have a potential race condition reported by a customer where somehow (even though it’s in a try/catch block), the `asmDefinition.Write (mergedAsmPath);` call is throwing in some cases due to a file access error (specifically file in use by another process).

There are a couple of possibilities:
 1. The file is actually still open by another msbuild invocation or another msbuild task running (either from our own tasks here, or some other part of the build process).
 2. Some previous process didn’t clean up the file completely, and it’s left in an open state.

I think 1 is most likely, since this seems to happen randomly and not always (and subsequent builds sometimes have the issue and sometimes do not).

This set of changes introduces a number of safeguards to mitigate the possibility of a race condition:

1. Using `IBuildEngine4` it registers the assembly being processed with the build engine for the given task so that other calls to the same task within the build engine (eg: building in parallel) can skip over processing for the assemblies already processed or in the middle of processing.
2. Also using the build engine global dictionary, the `AssemblyName` is cached for each assembly file read, so that subsequent calls can use the cached value instead of opening the assembly and reading it.
3. An exclusive read file lock is created on the assembly filename + ".lock".  each time the code intends to open the assembly for reading/writing, it will first attempt to obtain this exclusive lock before proceeding, which should ensure no other processes running our task should have the given assembly open already.
4. The build task now implements `ICancelableTask`.  This allows the wait on the exclusive read/file lock to be cancelled part way through the build (the wait can be as long as 60 seconds, so it's important this was cancelable).

**Proguard**
There was an issue with the proguard task where config files extracted and cached into the intermediate output dir were not being emitted to the FileWrites item group on subsequent incremental builds because the .stamp file was checked and then the task skipped if it existed.  This PR now makes the cached proguard config file paths saved to the .stamp file so subsequent incremental builds can read them from the .stamp file and emit them to the FileWrites item group.

Redth committed:
[xbd] add exclusive file lock check to cecil ops

We have a potential race condition reported by a customer where somehow (even though it’s in a try/catch block), the `asmDefinition.Write (mergedAsmPath);` call is throwing in some cases due to a file access error (specifically file in use by another process).

There are a couple of possibilities:
 1. The file is actually still open by another msbuild invocation or another msbuild task running (either from our own tasks here, or some other part of the build process).
 2. Some previous process didn’t clean up the file completely, and it’s left in an open state.

I think 1 is most likely, since this seems to happen randomly and not always (and subsequent builds sometimes have the issue and sometimes do not).

This change opens and waits for an exclusive file lock on the assembly name + “.dll” before it will do any reading or writing of the assembly.  This should ensure that our own task is no the culprit for the race condition.

Redth committed:
[xbd] use buildengine registered tasks

This should avoid even trying to read an assembly in cecil if another invocation of the build task has or is already processing the given assembly.

kzu commented on Util/Xamarin.Build.Download/source/Xamarin.Build.Download/BaseXamarinBuildResourceRestore.cs:
```diff
@@ -36,21 +37,37 @@ public abstract class BaseXamarinBuildResourceRestore : Task
 
+		protected IBuildEngine4 BuildEngine4 {
+			get {
+				return (IBuildEngine4)BuildEngine;
```
This might not work in xbuild...

kzu reviewed (approved):
Yeah, looks like this would do away with most of the concurrency issues. i can't think of a scenario where both the registered object *and* the lock wouldn't be sufficient... so 👍

Redth committed:
[xbd] Multiple file locking / caching strategies

This introduces caching around the use of `AssemblyName.GetAssemblyName()` calls so that we should never have to invoke that code more than once per assembly in a build.

We also use an exclusive file lock around the loading of the .dll assembly since there are other build tasks in XBD which open this assembly and lock the same file to not conflict.

Finally, we introduce a build engine caching to act as a lock around the processing of resources for an assembly so that multiple invocations for the same assembly will not try and do the same work over.

Redth committed:
[xbd] emit filewrites always in proguard step

The proguard config inclusion task was caching proguard config files and a .stamp file to the intermediate output directory.

However, when the .stamp file was checked on subsequent builds, it was not adding the previously cached proguard config files to the filewrites upon invocation, so it seems the build process was not aware of these, causing the dex files to be recompiled more often than necessary: https://github.com/xamarin/xamarin-android/issues/1256

This changes the build so that the .stamp file contents list the proguard config files that were previously cached, so they can be appended to the filewrites when the task runs again detecting the existing stamp file.

Redth commented:
Added to this PR is a fix for the proguard config task not always emitting the cached proguard config files from the intermediate output dir to the FileWrites, which made x.a rebuild dex files on incremental builds.

Redth committed:
[xbd] Preview nuget version

fuzzybinary commented:
What's the status on this PR? We're mostly interested in that progard fix so we can get better incremental build times.

bholmes commented:
The race condition is still a problem on Windows.  This is a top priority fix.  We will resume work on this issue later this week to release a fix asap.

jeromelaban commented:
@bholmes I've tested this fix for https://developercommunity.visualstudio.com/content/problem/190169/xamarinbuilddownload-047-has-a-race-condition.html and it seems to have fixed the mentioned concurrency issue. I do not have the proguard issue (not used in my context).
